core/nandtool/flashmng.c

```diff
@@ -0,0 +1,171 @@
+/*
+ * Phoenix-RTOS
+ *
+ * NANDtool utility
+ *
+ * Flash Server Manager
+ *
+ * Copyright 2019, 2021 Phoenix Systems
+ * Author: Hubert Buczynski, Lukasz Kosinski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <errno.h>
+#include <stdio.h>
+
+#include <sys/file.h>
+#include <sys/msg.h>
+
+#include "flashmng.h"
+
+
+static struct {
+	oid_t oid;
+	flashsrv_info_t info;
+} flashmng_common;
+
+
+int flashmng_writeraw(oid_t oid, unsigned int page, void *data, size_t size)
+{
+	msg_t msg = { 0 };
+	flash_i_devctl_t *idevctl = (flash_i_devctl_t *)msg.i.raw;
+	volatile flash_o_devctl_t *odevctl = (flash_o_devctl_t *)msg.o.raw;
+	flashsrv_info_t *info;
+	int err;
+
+	if ((info = flashmng_info(oid)) == NULL)
+		return -EFAULT;
+
+	msg.type = mtDevCtl;
+	msg.i.data = data;
+	msg.i.size = size;
+	idevctl->type = flashsrv_devctl_writeraw;
+	idevctl->write.oid = oid;
+	idevctl->write.address = page * (info->metasz + info->writesz);
+	idevctl->write.size = size;
+
+	if (((err = msgSend(oid.port, &msg)) < 0) || ((err = odevctl->err) < 0))
+		return err;
+
+	if (err != size)
+		return -EIO;
+
+	return EOK;
+}
+
+
+int flashmng_erase(oid_t oid, unsigned int start, unsigned int end)
+{
+	msg_t msg = { 0 };
+	flash_i_devctl_t *idevctl = (flash_i_devctl_t *)msg.i.raw;
+	volatile flash_o_devctl_t *odevctl = (flash_o_devctl_t *)msg.o.raw;
+	flashsrv_info_t *info;
+	int err;
+
+	if ((info = flashmng_info(oid)) == NULL)
+		return -EFAULT;
+
+	msg.type = mtDevCtl;
+	idevctl->type = flashsrv_devctl_erase;
+	idevctl->erase.oid = oid;
+	idevctl->erase.address = start * info->erasesz;
+	idevctl->erase.size = (end - start) * info->erasesz;
```

Comment:
earse all use case (end == 0?) not supported (FYI: this will be the most common usage path when using per-partition files to rewrite partition)

Reply:
Fixed (use `-e 0:0` option to erase whole device).